### PR TITLE
Custom item ordering and categorizing for the widget

### PR DESCRIPTION
**Pull request by karthikb351**
Implements #166.

- Customize the titles of rendered categories
- Enforce ordering of items

### Available via a new config `categories`
```javascript
window.Boxoffice.init({
        org: 'hasgeek', // Organisation to show
        itemCollection: 'd386dbcf-1a32-11e7-b315-5d5ee7239cc7', // ItemCollection ID to show
        paymentDesc: 'HasGeek', // Description for payments widget
        categories: [
                    {
                        'name': 'conference',
                        'title': 'Standard',
                        'item_ids': ['68e27823-1a33-11e7-842a-85544f6e9ad9']
                    },
                    {
                        'name': 'premium',
                        'title': 'Premium',
                        'item_ids': ['63d94e70-1b3a-11e7-85fb-737e61409825','19e47ba8-1e8e-11e7-8f39-45cbac6935ba']
                    }
                ]
        ]
      });
```

An Item's `name` is unique in an item collection, and is human readable so I chose that over `id`

---------
# Examples
## With categories
screen shot 2017-07-09 at 8 57 15 pm

```javascript
window.Boxoffice.init({
        org: 'hasgeek', // Organisation to show
        itemCollection: 'd386dbcf-1a32-11e7-b315-5d5ee7239cc7', // ItemCollection ID to show
        paymentDesc: 'HasGeek', // Description for payments widget
        categories: [
                    {
                        'name': 'conference',
                        'title': 'Standard',
                        'item_ids': ['68e27823-1a33-11e7-842a-85544f6e9ad9']
                    },
                    {
                        'name': 'premium',
                        'title': 'Premium',
                        'item_ids': ['63d94e70-1b3a-11e7-85fb-737e61409825','19e47ba8-1e8e-11e7-8f39-45cbac6935ba']
                    }
                ]
      });
```

## With duplicates and ordering
screen shot 2017-07-09 at 8 32 58 pm

```javascript
window.Boxoffice.init({
        org: 'hasgeek', // Organisation to show
        itemCollection: 'd386dbcf-1a32-11e7-b315-5d5ee7239cc7', // ItemCollection ID to show
        paymentDesc: 'HasGeek', // Description for payments widget
        categories: [
                    {
                        'name': 'conference',
                        'title': 'Standard',
                        'item_ids': ['68e27823-1a33-11e7-842a-85544f6e9ad9']
                    },
                    {
                        'name': 'premium',
                        'title': 'Premium',
                        'item_ids': ['63d94e70-1b3a-11e7-85fb-737e61409825','19e47ba8-1e8e-11e7-8f39-45cbac6935ba']
                    }
                ]
      });
```

## Without any value, defaults to server defined categories
screen shot 2017-07-09 at 8 33 22 pm

```javascript
window.Boxoffice.init({
        org: 'hasgeek', // Organisation to show
        itemCollection: 'd386dbcf-1a32-11e7-b315-5d5ee7239cc7', // ItemCollection ID to show
        paymentDesc: 'HasGeek' // Description for payments widget
            }
        ]
      });
```

### CODE REVIEW

**Comment by vidya-ram** [comment]
Wouldn't it better to do the filtering of items on the server side.

**Comment by vidya-ram** [comment]
Here customising category on the client side involves overriding the existing categories and also classification of items. Wouldn't this lead to confusion.

File: boxoffice/templates/boxoffice.js [hunk 1]
```diff
@@ -165,10 +166,22 @@ $(function() {
     }).done(function(data) {
       var lineItems = [];
 
+      // Flag: If items have been explicitly specified in the config
+      var should_filter_items = (widgetConfig.categories && widgetConfig.categories.length > 0 ? true : false);
```
Comment (shreyas-satish) [inline boxoffice/templates/boxoffice.js]: This line can just be `widgetConfig.categories.length > 0`.

**Comment by karthikb351** [comment]
@shreyas-satish Works with ids instead of names - I've updated the example here to reflect that. Good to merge?

**Pull request merged**